File: rewrite-java/src/test/java/org/openrewrite/java/style/ImportLayoutStyleTest.java
/*
 * Copyright 2020 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed, in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.java.style;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.module.paramnames.ParameterNamesModule;
import org.junit.jupiter.api.Test;
import org.openrewrite.Issue;
import org.openrewrite.config.DeclarativeNamedStyles;
import org.openrewrite.java.tree.*;
import org.openrewrite.marker.Markers;

import java.util.List;
import java.util.Set;

import static java.util.Collections.emptyList;
import static org.assertj.core.api.Assertions.assertThat;
import static org.openrewrite.Tree.randomId;

class ImportLayoutStyleTest {
    private final ObjectMapper mapper = new ObjectMapper()
            .registerModule(new ParameterNamesModule())
            .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);

    @Test
    void roundTripSerialize() throws Exception {
        var style = mapper.writeValueAsString(ImportLayoutStyle
                .builder()
                .packageToFold("java.awt.*")
                .packageToFold("java.swing.*", false)
                .staticPackageToFold("org.unit.Assert.*")
                .staticPackageToFold("org.mockito.Matchers.*", false)
                .importPackage("import java.*")
                .importPackage("import javax.*", false)
                .importAllOthers()
                .importStaticAllOthers()
                .build());

        var deserialized = mapper.readValue(style, ImportLayoutStyle.class);
        assertThat(style).isEqualTo(mapper.writeValueAsString(deserialized));
    }

    @Test
    void deserializeInDeclarativeNamedStyles() throws Exception {
        var style = new DeclarativeNamedStyles(
                randomId(),
                "name",
                "displayName",
                "description",
                Set.of("tag1", "tag2"),
                List.of(ImportLayoutStyle.builder()
                        .classCountToUseStarImport(5)
                        .nameCountToUseStarImport(5)
                        .importPackage("java.*")
                        .blankLine()
                        .importPackage("javax.*")
                        .blankLine()
                        .importAllOthers()
                        .blankLine()
                        .importPackage("org.springframework.*")
                        .blankLine()
                        .importStaticAllOthers()
                        .packageToFold("java.awt.*")
                        .packageToFold("java.swing.*")
                        .build()
                )
        );

        mapper.readValue(mapper.writeValueAsBytes(style), DeclarativeNamedStyles.class);
    }

    @Test
    @Issue("https://github.com/openrewrite/rewrite/issues/4196")
    void addImportInPresenceOfDuplicateOtherImport() {
        ImportLayoutStyle style = new ImportLayoutStyle(
                Integer.MAX_VALUE, Integer.MAX_VALUE, emptyList(), emptyList());
        JRightPadded<J.Import> import1 = new JRightPadded<>(
                new J.Import(
                        randomId(),
                        Space.EMPTY,
                        Markers.EMPTY,
                        new JLeftPadded<>(Space.SINGLE_SPACE, true, Markers.EMPTY),
                        TypeTree.build("pkg.Clazz.MEMBER_1").withPrefix(Space.SINGLE_SPACE),
                        null),
                Space.EMPTY,
                Markers.EMPTY);
        JRightPadded<J.Import> import2 = new JRightPadded<>(
                new J.Import(
                        randomId(),
                        Space.EMPTY,
                        Markers.EMPTY,
                        new JLeftPadded<>(Space.SINGLE_SPACE, true, Markers.EMPTY),
                        TypeTree.build("pkg.Clazz.MEMBER_1").withPrefix(Space.SINGLE_SPACE),
                        null),
                Space.EMPTY,
                Markers.EMPTY);
        J.Import importToAdd = new J.Import(
                randomId(),
                Space.EMPTY,
                Markers.EMPTY,
                new JLeftPadded<>(Space.SINGLE_SPACE, true, Markers.EMPTY),
                TypeTree.build("pkg.Clazz.MEMBER_2").withPrefix(Space.SINGLE_SPACE),
            null);
        assertThat(style.addImport(List.of(import1, import2), importToAdd, null, emptyList()))
                .containsExactlyInAnyOrder(
                        import1, import1, new JRightPadded<>(importToAdd, Space.EMPTY, Markers.EMPTY));
    }

    @Test
    @Issue("https://github.com/openrewrite/rewrite/issues/4241")
    void addImportWithNewLineInUnsortedImportList() {
        ImportLayoutStyle style = new ImportLayoutStyle(
          Integer.MAX_VALUE, Integer.MAX_VALUE, emptyList(), emptyList());
        JRightPadded<J.Import> import0 = new JRightPadded<>(
          new J.Import(
            randomId(),
            Space.EMPTY,
            Markers.EMPTY,
            new JLeftPadded<>(Space.SINGLE_SPACE, true, Markers.EMPTY),
            TypeTree.build("pkg.AAA.MEMBER_0").withPrefix(Space.SINGLE_SPACE),
            null),
          Space.EMPTY,
          Markers.EMPTY);
        JRightPadded<J.Import> import1 = new JRightPadded<>(
          new J.Import(
            randomId(),
            Space.EMPTY,
            Markers.EMPTY,
            new JLeftPadded<>(Space.SINGLE_SPACE, true, Markers.EMPTY),
            TypeTree.build("pkg.Clazz.MEMBER_3").withPrefix(Space.SINGLE_SPACE),
            null),
          Space.EMPTY,
          Markers.EMPTY);
        JRightPadded<J.Import> import3 = new JRightPadded<>(
          new J.Import(
            randomId(),
            Space.EMPTY,
            Markers.EMPTY,
            new JLeftPadded<>(Space.SINGLE_SPACE, true, Markers.EMPTY),
            TypeTree.build("pkg.Clazz.MEMBER_1").withPrefix(Space.SINGLE_SPACE),
            null),
          Space.EMPTY,
          Markers.EMPTY);
        J.Import importToAdd = new J.Import(
          randomId(),
          Space.EMPTY,
          Markers.EMPTY,
          new JLeftPadded<>(Space.EMPTY, true, Markers.EMPTY),
          TypeTree.build("pkg.Clazz.MEMBER_2").withPrefix(Space.SINGLE_SPACE),
          null);

        assertThat(style.addImport(List.of(import0, import1, import3), importToAdd, null, emptyList()).get(1).getElement().getPrefix()).isEqualTo(Space.format("\n"));
    }
}

File: rewrite-java/src/test/kotlin/org/openrewrite/java/style/ImportLayoutStyleTest.kt
/*
 * Copyright 2020 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.java.style

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.paramnames.ParameterNamesModule
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.openrewrite.Tree.randomId
import org.openrewrite.config.DeclarativeNamedStyles
import org.openrewrite.style.Style

class ImportLayoutStyleTest {
    companion object {
        private val mapper = ObjectMapper()
                .registerModule(ParameterNamesModule())
                .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
    }

    @Test
    fun roundTripSerialize() {
        val style = mapper.writeValueAsString(ImportLayoutStyle
            .builder()
            .packageToFold("java.awt.*")
            .packageToFold("java.swing.*", false)
            .staticPackageToFold("org.unit.Assert.*")
            .staticPackageToFold("org.mockito.Matchers.*", false)
            .importPackage("import java.*")
            .importPackage("import javax.*", false)
            .importAllOthers()
            .importStaticAllOthers()
            .build())

        val deserialized = mapper.readValue(style, ImportLayoutStyle::class.java)
        assertThat(style).isEqualTo(mapper.writeValueAsString(deserialized))
    }

    @Test
    fun deserializeStyle() {
        val styleConfig = mapOf(
                "@c" to ImportLayoutStyle::class.qualifiedName,
                "@ref" to 1,
                "classCountToUseStarImport" to 999,
                "nameCountToUseStarImport" to 998,
                "layout" to listOf(
                        "import java.*",
                        "<blank line>",
                        "import javax.* without subpackages",
                        "<blank line>",
                        "import all other imports",
                        "<blank line>",
                        "import org.springframework.*",
                        "<blank line>",
                        "import static all other imports"
                ),
                "packagesToFold" to listOf(
                    "import java.awt.*",
                    "import java.swing.* without subpackages",
                    "import static org.unit.Assert.*",
                    "import static org.mockito.Matchers.* without subpackages"
                )
        )

        val style = mapper.convertValue(styleConfig, ImportLayoutStyle::class.java)
        assertThat(style.classCountToUseStarImport).isEqualTo(999)
        assertThat(style.nameCountToUseStarImport).isEqualTo(998)

        // round trip
        when (val importLayout: Style = mapper.readValue(mapper.writeValueAsBytes(style), Style::class.java)) {
            is ImportLayoutStyle -> {
                assertThat(importLayout.classCountToUseStarImport).isEqualTo(999)
                assertThat(importLayout.nameCountToUseStarImport).isEqualTo(998)
                assertThat(importLayout.layout.size).isEqualTo(9)

                assertThat(importLayout.layout[0])
                    .isInstanceOf(ImportLayoutStyle.Block.ImportPackage::class.java)
                    .matches { b -> !(b as ImportLayoutStyle.Block.ImportPackage).isStatic }
                    .matches { b -> (b as ImportLayoutStyle.Block.ImportPackage).packageWildcard.toString() == "java\\..+" }

                assertThat(importLayout.layout[1]).isInstanceOf(ImportLayoutStyle.Block.BlankLines::class.java)

                assertThat(importLayout.layout[2])
                    .isInstanceOf(ImportLayoutStyle.Block.ImportPackage::class.java)
                    .matches { b -> !(b as ImportLayoutStyle.Block.ImportPackage).isStatic }
                    .matches { b -> (b as ImportLayoutStyle.Block.ImportPackage).packageWildcard.toString() == "javax\\.[^.]+" }

                assertThat(importLayout.layout[3]).isInstanceOf(ImportLayoutStyle.Block.BlankLines::class.java)

                assertThat(importLayout.layout[4])
                    .isInstanceOf(ImportLayoutStyle.Block.ImportPackage::class.java)
                    .matches { b -> !(b as ImportLayoutStyle.Block.ImportPackage).isStatic }
                    .matches { b -> (b as ImportLayoutStyle.Block.ImportPackage).packageWildcard.toString() == ".+" }

                assertThat(importLayout.layout[5]).isInstanceOf(ImportLayoutStyle.Block.BlankLines::class.java)

                assertThat(importLayout.layout[6])
                    .isInstanceOf(ImportLayoutStyle.Block.ImportPackage::class.java)
                    .matches { b -> !(b as ImportLayoutStyle.Block.ImportPackage).isStatic }
                    .matches { b -> (b as ImportLayoutStyle.Block.ImportPackage).packageWildcard.toString() == "org\\.springframework\\..+" }

                assertThat(importLayout.layout[7]).isInstanceOf(ImportLayoutStyle.Block.BlankLines::class.java)

                assertThat(importLayout.layout[8])
                    .isInstanceOf(ImportLayoutStyle.Block.ImportPackage::class.java)
                    .matches { b -> (b as ImportLayoutStyle.Block.ImportPackage).isStatic }
                    .matches { b -> (b as ImportLayoutStyle.Block.ImportPackage).packageWildcard.toString() == ".+" }

                assertThat(importLayout.packagesToFold[0])
                    .isInstanceOf(ImportLayoutStyle.Block.ImportPackage::class.java)
                    .matches { b -> !(b as ImportLayoutStyle.Block.ImportPackage).isStatic }
                    .matches { b -> (b as ImportLayoutStyle.Block.ImportPackage).packageWildcard.toString() == "java\\.awt\\..+" }

                assertThat(importLayout.packagesToFold[1])
                    .isInstanceOf(ImportLayoutStyle.Block.ImportPackage::class.java)
                    .matches { b -> !(b as ImportLayoutStyle.Block.ImportPackage).isStatic }
                    .matches { b -> (b as ImportLayoutStyle.Block.ImportPackage).packageWildcard.toString() == "java\\.swing\\.[^.]+" }

                assertThat(importLayout.packagesToFold[2])
                    .isInstanceOf(ImportLayoutStyle.Block.ImportPackage::class.java)
                    .matches { b -> (b as ImportLayoutStyle.Block.ImportPackage).isStatic }
                    .matches { b -> (b as ImportLayoutStyle.Block.ImportPackage).packageWildcard.toString() == "org\\.unit\\.Assert\\..+" }

                assertThat(importLayout.packagesToFold[3])
                    .isInstanceOf(ImportLayoutStyle.Block.ImportPackage::class.java)
                    .matches { b -> (b as ImportLayoutStyle.Block.ImportPackage).isStatic }
                    .matches { b -> (b as ImportLayoutStyle.Block.ImportPackage).packageWildcard.toString() == "org\\.mockito\\.Matchers\\.[^.]+" }
            }
        }
    }

    @Test
    fun deserializeInDeclarativeNamedStyles() {
        val style = DeclarativeNamedStyles(
                randomId(),
                "name",
                "displayName",
                "description",
                setOf("tag1", "tag2"),
                listOf<Style>(ImportLayoutStyle.builder()
                        .classCountToUseStarImport(5)
                        .nameCountToUseStarImport(5)
                        .importPackage("java.*")
                        .blankLine()
                        .importPackage("javax.*")
                        .blankLine()
                        .importAllOthers()
                        .blankLine()
                        .importPackage("org.springframework.*")
                        .blankLine()
                        .importStaticAllOthers()
                        .packageToFold("java.awt.*")
                        .packageToFold("java.swing.*")
                        .build()
                )
        )
        mapper.readValue(mapper.writeValueAsBytes(style),
                DeclarativeNamedStyles::class.java)
    }
}
